File: Linked List/Doubly linkedlist/LinkedList.cpp
#include<bits/stdc++.h>

using namespace std;

class Node{
    public:
        Node *prev_node;
        int data;
        Node *next_node;
    
        Node(){prev_node = NULL; next_node = NULL;}
        Node(int data);
        
        Node* get_next_node();
        Node* get_prev_node();
        int get_data();

        void set_next_node(Node*);
        void set_prev_node(Node*);
        void set_data(int);
};

Node :: Node(int data){this->data = data;}

Node* Node :: get_next_node(){return next_node;}

Node* Node :: get_prev_node(){return prev_node;}

int Node :: get_data(){return data;}

void Node :: set_next_node(Node *next_node){this->next_node = next_node;}

void Node :: set_prev_node(Node *prev_node){this->prev_node = prev_node;}

void Node :: set_data(int data){this->data = data;}

class LinkList{
    public:
        Node *head,*tail;

        int length;

        LinkList() {
            head = tail = NULL;
            length = 0;
        }

        bool empty();

        void push_front(int);

        void push_back(int);

        void insert(int, int);

        int pop_front();

        int pop_back();

	    int remove(int);

        bool deletion(int);

        void print();

        void print_reverse();
};

bool LinkList ::empty(){
    return (head == NULL || tail == NULL);
}

void LinkList :: push_front(int data){
    Node *new_node = new Node(data);
    if(empty()){
        head = tail = new_node;
        return;
    }
    new_node->set_next_node(head);
    head->set_prev_node(new_node);
    length++;
    head = new_node;
}

void LinkList :: push_back(int data){
    Node *new_node = new Node(data);
    if(empty()){
        head = tail = new_node;
        return;
    }
    new_node->set_prev_node(tail);
    tail->set_next_node(new_node);
    length++;
    tail = new_node;
}

void LinkList :: insert(int data, int position){
    if(position > length){
        cout<< "Sorry your position is much big as length of list" << endl;
        return;
    }
    if(position == 0){
        push_front(data);
        return;
    }
    if(position == length+1){
        push_back(data);
        return;
    }
    Node *new_node = new Node(data);
    Node *temp_node = head;
    int i=0;
    while(i<position-1){
        temp_node = temp_node->get_next_node();
        i++;
    }
    new_node->set_next_node(temp_node->get_next_node());
    new_node->set_prev_node(temp_node);

    temp_node->get_next_node()->set_prev_node(new_node);
    temp_node->set_next_node(new_node);
}

int LinkList :: pop_front(){
    if(empty()){
        cout << "Your list is empty" << endl;
        return -1;
    }
    if(head->get_next_node() == NULL)
    {
        int data = head->get_data();
        head = tail = NULL;
        return data;
    }
    Node *temp_node = head;
    head = temp_node->get_next_node();
    head->set_prev_node(NULL);
    int data = temp_node->get_data();
    delete(temp_node);
    length--;
    return data;
}

int LinkList :: pop_back(){
    if(empty()){
        cout << "Your list is empty" << endl;
        return -1;
    }
    if(tail->get_prev_node() == NULL)
    {
        int data = tail->get_data();
        head = tail = NULL;
        return data;
    }
    Node *temp_node = tail;
    tail = temp_node->get_prev_node();
    tail->set_next_node(NULL);
    int data = temp_node->get_data();
    delete(temp_node);
    length--;
    return data;
}

int LinkList :: remove(int position){
    if(empty()){
        cout << "Your List is Empty" <<endl;
        return -1;
    }
    if(position > length || position < 0){
        cout<<"Enter valid Position"<<endl;
        return -1;
    }
    if(position == 0)   return pop_front();

    if(position == length)   return pop_back();

    int i = 0;
    Node* temp_node = head;
    while(i<position-1){
        temp_node = temp_node->get_next_node();
        i++;
    }
    Node *temp = temp_node->get_next_node();
    temp_node->set_next_node(temp->get_next_node());
    temp->get_next_node()->set_prev_node(temp_node);

    int data = temp->data;
    delete(temp);
    length--;
    return data;
}

bool LinkList :: deletion(int num){
    if(empty()){
        cout << "Your List is Empty" <<endl;
        return false;
    }
    if(head->get_data() == num){
        pop_front();
        return true;
    }
    Node *temp_node = head;
    Node *temp = temp_node->get_next_node();
    while(temp->get_data() != num && temp->get_next_node() != NULL){
        temp = temp->get_next_node();
        temp_node = temp_node->get_next_node();
    }
    if(temp->get_data() == num){
        temp_node->set_next_node(temp->get_next_node());
        temp->get_next_node()->set_prev_node(temp_node);
        delete(temp);
        length--;
        return true;
    }
    return false;
}

void LinkList :: print(){
    if(empty()){
        cout<<"Your list is empty"<<endl;
        return;
    }
    cout << "Your entered list is from head : NULL <=> ";
    Node *temp_node = head;
    while(temp_node != NULL){
        cout << temp_node->get_data() << " <=> ";
        temp_node = temp_node->get_next_node();
    }
        cout << "NULL" << endl;
}

void LinkList :: print_reverse(){
    if(empty()){
        cout<<"Your list is empty"<<endl;
        return;
    }
    cout << "Your entered list is from tail : NULL <=> ";
    Node *temp_node = tail;
    while(temp_node != NULL){
        cout << temp_node->get_data() << " <=> ";
        temp_node = temp_node->get_prev_node();
    }
        cout << "NULL" << endl;
}


int main(){
    LinkList l;
    l.push_front(2);
    l.push_front(1);
    l.push_front(0);
    l.push_back(3);
    l.push_back(4);
    l.push_back(5);
    l.push_back(6);
    l.insert(1,1);
    cout << "length is " << l.length << endl;
    l.print();
    l.print_reverse();
    cout << "removing index 1 is " << l.deletion(1) << endl;
    cout << "length is " << l.length << endl;
    l.print();
    l.print_reverse();
}

File: Linked List/Doubly linkedlist/LinkedList.java
// A Node class this create as a double linkedList
class Node{

    private Node prevNode; // this is Node which store prev Node's address
    private int data; // it store data
    private Node nextNode; // this is a node which store next Node's address;
    
    // Node constructer ---------------------

    public Node(){
        // when we create a new node so we make their previous and next node with null
        this.prevNode = null;
        this.nextNode = null;
    }

    public Node(int data){
        // it is constructer overloading which take input data value of created new node
        this.data = data;
        this.prevNode = null;
        this.nextNode = null;
    }

    // Function Decleration ------------------

    // it tell what data of that node
    public int getData() { return data; }

    // it tell which node is next
    public Node getNextNode() { return nextNode; }

    // it tell what node is pre of this node
    public Node getPrevNode() { return prevNode; }

    // it set data inside a node
    public void setData(int data) { this.data = data; }

    // it set previous node
    public void setNextNode(Node node) { this.nextNode = node; }

    // it set the next node
    public void setPrevNode(Node node) { this.prevNode = node; }
}

class LinkedList{
    // create a two node one for head and another for tail
    Node head,tail;
    int length;

    // constructer
    public LinkedList(){
        head = tail = null; // make both head and tail null when object is creted
        length=0; // assign length with zero
    }

    // check a list is empty or not
    public boolean isEmpty(){
        return (head == null || tail == null);
    }


    // insertion

    // enter data at front of list
    public void pushFront(int data){
        Node newNode = new Node(data);
        // checking node is empty or not
        if(isEmpty()){
            head = tail = newNode;
            length++;
            return;
        }
        // making connction from new note to all node at begining
        newNode.setNextNode(head);
        head.setPrevNode(newNode);
        length++;
        head = newNode;
    }

    // enter data at last of list
    public void pushBack(int data){
        Node newNode = new Node(data);
        // checking list is empty or not
        if(isEmpty()){
            head = tail = newNode;
            length++;
            return;
        }
        // make connection from new node to all rest node at last
        newNode.setPrevNode(tail);
        tail.setNextNode(newNode);
        length++;
        tail = newNode;
    }

    // this function is used to insert data at specific position
    public void insert(int data, int position){
        if(isEmpty()){
            System.out.println("List is empty");
            return;
        }
        // check given position is valid or not
        if(position > length || position < 0){
            System.out.println("Enter position well fromat");
            return;
        }
        
        if(position == 0){
            pushFront(data);
            return;
        }

        if(position == length){
            pushBack(data);
        }
        /*
            Logic is that first run loop from head to tail
            and check postion is specific postion where 
            i have to store data so i keep a addres of 
            previous and nexd of that node and put node their and 
            make connection with both node
        */
        int i = 0;
        Node tempNode = head;
        Node temp = tempNode.getNextNode();
        // travel to specific position
        while(i<position-1){
            temp = temp.getNextNode();
            tempNode = tempNode.getNextNode();
            i++;
        }

        // create a new node and make connection with it
        Node newNode = new Node(data);
        tempNode.setNextNode(newNode);
        newNode.setNextNode(temp);
        newNode.setPrevNode(tempNode);
        temp.setPrevNode(newNode);
    }


    // Deletion

    // to delete a data from front
    public int popFront(){
        // check list is empty or not
        if(isEmpty()){
            System.out.println("List is Empty");
            return -1;
        }

        // if list is only one element so make head and tali null 
        Node temp = head;
        if(head == tail){
            head = tail = null;
            length--;
            return temp.getData();
        }    
        
        // otherwise we remove link from first node to another node 
        head = head.getNextNode();
        head.setPrevNode(null);
        length--;
        return temp.getData();
    }

    // to delete from back
    public int popBack(){
        // check list is empty or not
        if(isEmpty()){
            System.out.println("Your List is Empty...");
            return -1;
        }
        
        // if list is only one element so make head and tali null 
        Node temp = tail;
        if(head == tail){
            head = tail = null;
            return temp.getData();
        }

        // otherwise we remove link from last node to another node
        tail = tail.getPrevNode();
        tail.setNextNode(null);
        length--;
        return temp.getData();
    }

    public int delete(int position){
        // check list is empty or not
        if(isEmpty()){
            System.out.println("Your List is Empty...");
            return -1;
        }
        // check position value is not a inlegal
        if(position < 0 || position > length){
            System.out.println("Enter position well fromat");
            return -1;
        }
        
        if(position == 0){
            return popFront();
        }

        if(position == length){
            return popBack();
        }

        // got to that which one want to delete and remove it
        int i = 0;
        Node tempNode = head;
        Node temp = tempNode.getNextNode();
        while(i<position-1){
            tempNode = tempNode.getNextNode();
            temp = temp.getNextNode();
            i++;
        }
        int data = temp.getData();
        temp = temp.getNextNode();
        tempNode.setNextNode(temp);
        temp.setPrevNode(tempNode);
        return data;
    }
    
     // traversal
    public void print(){
        if(isEmpty()){
            System.out.println("Your List is empty");
            return;
        }

        Node tempNode = head;
        System.out.print("Yor inserted List is from head : NULL <=> ");
        while(tempNode != null){
            System.out.print(tempNode.getData() + " <=> ");
            tempNode = tempNode.getNextNode();
        }
        System.out.print("NULL\n");
    }

    public void printReverse(){
        if(isEmpty()){
            System.out.println("Your List is empty");
            return;
        }

        Node tempNode = tail;
        System.out.print("Yor inserted List is from tail : NULL <=> ");
        while(tempNode != null){
            System.out.print(tempNode.getData() + " <=> ");
            tempNode = tempNode.getPrevNode();
        }
        System.out.print("NULL\n");
    }


    public int find(int num){
        if(isEmpty()){
            System.out.println("Your List is empty");
            return -1;
        }

        int i=0,j=length;
        Node start = head,end = tail;
        while(i != j){
            if(start.getData() == num){
                return i;
            }
            i++;
            start = start.getNextNode();
            if(end.getData() == num){
                return j;
            }
            j--;
            end = end.getPrevNode();
        }
        
        if(start.getData() == num)
            return i;

        return -1;
    }



    public static void main(String []arg){
        LinkedList l = new LinkedList();
        l.pushBack(4);
        l.pushBack(5);
        l.pushBack(6);
        l.pushFront(3);
        l.pushFront(2);
        l.pushFront(1);
        l.pushBack(7);
        l.pushBack(8);
        l.pushBack(9);
        l.pushFront(0);
        l.print();
        l.printReverse();
        System.out.println("Data is inserted in list at position 8: ");
        l.insert(8,8);
        l.print();
        l.printReverse();
        System.out.println("Length of length is : "+ l.length);
        System.out.println("Data deleted at position from 8 is: "+l.delete(8));
        l.print();
        l.printReverse();
        if(l.find(9) != -1)
            System.out.println("9 is avilable at position: "+l.find(9));
        else System.out.println("9 is n0t avilable in list");
    }
}